### PR TITLE
Add meta-review to gamification

### PR DESCRIPTION
**Pull request by shikharvaish28**
From this PR, we add the meta-review activity on the gamification page of the active newcomers.

Closes https://github.com/coala/community/issues/181

### CODE REVIEW

**Comment by shikharvaish28** [comment]
Netlify deploy is working on my end of another branch but not on this one. :(

File: gamification/process/update.py [hunk 1]
```diff
@@ -56,7 +57,7 @@ def update_participants_data_with_mr(mr):
 
         # Update participant data
         mr_author.add_points(mr_points,
-                             mr_activity_string,
+                             mr_activity_string,  # Ignore PyFlakesBear
```
Comment (li-boxuan) [inline gamification/process/update.py]: Any reason for this change?

File: gamification/process/update.py [hunk 2]
```diff
@@ -135,6 +136,29 @@ def get_participant_objects():
     return participants
 
 
+def update_participants_data_with_meta_review(participants):
```
Comment (li-boxuan) [inline gamification/process/update.py]: On line 35 (https://github.com/coala/community/pull/190/files#diff-0bee1776ac8d509157f197b459b8b9afR35) you are calling `update_participants_data_with_meta_review(participant)` but here it is called `participants`. So should the argument be one participant or many participants?

File: gamification/process/update.py [hunk 3]
```diff
@@ -135,6 +136,29 @@ def get_participant_objects():
...
+    """
+    pos_in_meta_review = meta_review_participants.objects.get(
+                        pos_in=participants.pos_in)
+    pos_out_meta_review = meta_review_participants.objects.get(
+                        pos_out=participants.pos_out)
+    neg_in_meta_review = meta_review_participants.objects.get(
+                        neg_in=participants.neg_in)
+    neg_out_meta_review = meta_review_participants.objects.get(
+                        neg_out=participants.neg_out)
+    if pos_in_meta_review > 0 or pos_out_meta_review > 0 or
```
Comment (li-boxuan) [inline gamification/process/update.py]: See https://stackoverflow.com/questions/181530/styling-multi-line-conditions-in-if-statements

File: gamification/process/update.py [hunk 4]
```diff
@@ -135,6 +136,29 @@ def get_participant_objects():
...
+def update_participants_data_with_meta_review(participants):
+    """
+    Update participants based on meta-review
+
+    This method updates every participant based on the meta-review
+    recieved or given. If the meta-review is completed then it calls
+    'Completed Meta-Review' activity.
+    """
+    pos_in_meta_review = meta_review_participants.objects.get(
+                        pos_in=participants.pos_in)
```
Comment (li-boxuan) [inline gamification/process/update.py]: This part makes me a bit confused. Does this ever work? Apparently `participants` come from `gamification.models`, which doesn't have `pos_in` field.

`meta_review_participants.objects.get(pos_in=some number)` doesn't make sense to me either. This returns an object whose `pos_in` field = `some number`.

File: .gitignore [hunk 5]
```diff
@@ -488,3 +488,5 @@ _site/
 
 # Pytest profile files
 prof/
+*.orig
+.vscode/settings.json
```
Comment (li-boxuan) [inline .gitignore]: Create a new issue for changes you want to make that are unrelated to this pr.
Also, `.vscode` already exists in `.gitignore`.

File: gamification/process/update.py (hunk 2 above)
Comment (sks444) [inline gamification/process/update.py]: One idea is that this method would take list of all the participants objects from the
meta-review partcipant model...and then iterate through each of the participant object from the list while filtering out only those participant whose usernames are in [active_newcomer_list](https://github.com/coala/community/blob/master/data/newcomers.py#L9)(as gamifiction system only work for those participant) and then if that participant have `completed meta-review` then get the object of that participant from the gamification participant model and add the meta-review activity for that participant.

Comment (sks444) [reply gamification/process/update.py]: I think maybe you're getting confused b/w `meta-review Participants` and `gamification Participants`, meta-review system works for all members at coala and gamification system works for only [active newcomers](https://github.com/coala/community/blob/master/data/newcomers.py#L9) at coala.

participants from `get_participant_objects()` are coming from gamification models, so they do not have fields like `pos_in` and `neg_out` etc..so you can't do[1](https://github.com/coala/community/pull/190/files#diff-49eb4e29b7797643ccd41983aeb4c575R148) as @li-boxuan mentioned there.

and where is `meta_review_completed` field? if you want to do something like `participant.meta_review_completed` with participant object you need a field of that name to the `Participant` model. I think you don't need to do something like that, when you find that a participant has completed meta-review, just add the `meta-review completed` activity for that participant by using [`add_activity`](https://github.com/coala/community/blob/master/gamification/models.py#L135) method.

Maybe rebase this commit on top of https://github.com/coala/community/pull/152 so that this pr passes CI and we can see you changes running.
> I'm testing these changes on another branch so that while pushing the changes, my PR doesn't gets bad with Travis test failing messages

File: .gitignore (hunk 5 above)
Comment (shikharvaish28) [reply .gitignore]: Created https://github.com/coala/community/issues/194

File: .gitignore [hunk 6]
```diff
@@ -488,3 +488,4 @@ _site/
 
 # Pytest profile files
 prof/
+*.orig
```
Comment (li-boxuan) [inline .gitignore]: Remove this

File: gamification/data/config.py [hunk 7]
```diff
@@ -76,6 +76,9 @@ def create_badge_activities():
             name='Solved a area/genericbears bear proposal issue'),
         BadgeActivity(
             name='Solved a area/lintbears bear proposal issue'),
+        BadgeActivity(
+            name='Completed Meta-Review'
+        ),
```
Comment (li-boxuan) [inline gamification/data/config.py]: IMO it's better to keep the format consistent, i.e. put `),` at the end of last line

File: gamification/management/commands/update_participants_data.py [hunk 8]
```diff
@@ -31,3 +33,5 @@ def handle(self, *args, **options):
             update_participants_data_with_issue(issue)
         for participant in get_participant_objects():
             award_badges(participant)
+        for participant in get_participant_objects_meta_review():
```
Comment (li-boxuan) [inline gamification/management/commands/update_participants_data.py]: IMO `get_meta_review_participant_objects` sounds better than `get_participant_objects_meta_review`

File: gamification/process/update.py [hunk 9]
```diff
@@ -135,6 +136,34 @@ def get_participant_objects():
     return participants
 
 
+def get_participant_objects_meta_review():
+    meta_review_participant = meta_review_participants.objects.all()
+    return meta_review_participant
```
Comment (li-boxuan) [inline gamification/process/update.py]: Just simply do `return meta_review_participants.objects.all()` is enough

File: gamification/process/update.py [hunk 10]
```diff
@@ -135,6 +136,34 @@ def get_participant_objects():
...
+                        pos_in=meta_review_participants.pos_in)
+    pos_out_meta_review = meta_review_participants.objects.get(
+                        pos_out=meta_review_participants.pos_out)
+    neg_in_meta_review = meta_review_participants.objects.get(
+                        neg_in=meta_review_participants.neg_in)
+    neg_out_meta_review = meta_review_participants.objects.get(
+                        neg_out=meta_review_participants.neg_out)
+    if pos_in_meta_review > 0 or pos_out_meta_review > 0 or \
+       neg_in_meta_review > 0 or neg_out_meta_review > 0:
+        participant.meta_review_completed = True
```
Comment (li-boxuan) [inline gamification/process/update.py]: Is this field defined anywhere? Moreover, IMO this field should belong to gamification participant, not meta-review participant. Seems you are passing a meta-review participant to this function `update_participants_data_with_meta_review`

File: gamification/process/update.py [hunk 11]
```diff
@@ -135,6 +136,34 @@ def get_participant_objects():
...
+def update_participants_data_with_meta_review(participant):
+    """
+    Update participants based on meta-review
+
+    This method updates every participant based on the meta-review
+    received or given. If the meta-review is completed then it calls
+    'Completed Meta-Review' activity.
+    """
+    pos_in_meta_review = meta_review_participants.objects.get(
+                        pos_in=meta_review_participants.pos_in)
```
Comment (li-boxuan) [inline gamification/process/update.py]: I don't quite understand this. If you want to get the value of `pos_in` of the current meta-review participant (seems the parameter of this function is meta_review participant), you should do `pos_in_meta_review = participant.pos_in`

**Comment by shikharvaish28** [comment]
I'm passing a meta-review participant from a list of participants extracted from ```get_meta_review_participant_objects```. Then, we iterate over that list and send each member to function ```update_participants_data_with_meta_review``` since the meta-review is done for each participant. If ```pos_in``` and other attributes are found to be positive then we call [add_activity](https://github.com/coala/community/blob/master/gamification/models.py#L135) to update the activity.

@li-boxuan @sks444 Your thoughts please.

File: meta_review/models.py [hunk 12]
```diff
@@ -95,3 +95,12 @@ class Reaction(models.Model):
     giver = models.ForeignKey(Participant, related_name='give', null=True)
     receiver = models.ForeignKey(Participant, related_name='receive', null=True)
     review = models.ForeignKey(Comment, null=True)
+
+    def meta_review_add_activity(self):
```
Comment (li-boxuan) [inline meta_review/models.py]: meta-review activity shouldn't be defined in the meta-review app, but in the gamification app.

Comment (shikharvaish28) [reply meta_review/models.py]: Will change it. Apart from that is everything fine?

Comment (li-boxuan) [reply meta_review/models.py]: I think this is the major problem. Once the logic gets correct then we can focus on minor issues like coding style, etc.

File: gamification/models.py [hunk 13]
```diff
@@ -186,3 +187,13 @@ def add_badges(self, activities):
         badges = self.find_badges_for_activity(activities)
         for badge in badges:
             self.badges.add(badge)
+
+
+def meta_review_add_activity(self):
+    """
+    Add a meta-review activity to the participant
+    """
+    activity = Reaction.object.create(
```
Comment (li-boxuan) [inline gamification/models.py]: This part does not make sense to me. I don't understand what you are doing with `Reaction` here. Once you know if a participant has finished the meta-review process, the rest staff has nothing to do with the meta-review app.

Comment (li-boxuan) [reply gamification/models.py]: What you said at https://github.com/coala/community/pull/190#issuecomment-423495158 looks good. However, you are not handling `update the activity` correctly.

Comment (shikharvaish28) [reply gamification/models.py]: > This part does not make sense to me. I don't understand what you are doing with `Reaction` here. Once you know if a participant has finished the meta-review process, the rest staff has nothing to do with the meta-review app.

In ```meta_review_add_activity ```, I'm using the details from```Reaction``` class of ```meta-review``` app to add the ```Completed Meta-Review``` activity (In the same way as that of ```add_activity``` in the ```gamification```.



> What you said at [#190 (comment)](https://github.com/coala/community/pull/190#issuecomment-423495158) looks good. However, you are not handling `update the activity` correctly here.
> 
> Moreover, this should not be a standalone function. At least make it a method of `Participant` class in `gamification/models.py`. Maybe we don't even need it - you can simply put the logic in `update_participants_data_with_meta_review` function. After you implement the correct logic, we can start talking about coding style.

I added this as a function because of the pattern followed by ```add_activity``` of ```gamification```.

If you feel that I should remove this function as a whole and implement the logic in ```update_participants_data_with_meta_review```, I can do it.

Comment (li-boxuan) [reply gamification/models.py]: > I'm using the details fromReaction class of meta-review app to add the Completed Meta-Review activity

An activity should be an instance of `Activity` class. When you do `Reaction.object.create` you are creating an instance of `Reaction`, which does not make sense (and `Reaction` does not have `activity` field btw).

> I added this as a function because of the pattern followed by add_activity of gamification.

Okay, I didn't notice that point ;) Then don't put it in `models.py`. But we can put this thing aside and make the logic go correctly first.

**Comment by shikharvaish28** [comment]
> https://gci-leaders.netlify.com/data.json

The link is working and there's no issue on my local machine too while running ```coala``` .  Why is it failing here?

File: gamification/process/update.py [hunk 14]
```diff
@@ -135,6 +136,27 @@ def get_participant_objects():
...
+    received or given. If the meta-review is completed then it calls
+    'Completed Meta-Review' activity.
+    """
+    pos_in_meta_review = participant.pos_in
+    pos_out_meta_review = participant.pos_out
+    neg_in_meta_review = participant.neg_in
+    neg_out_meta_review = participant.neg_out
+    if pos_in_meta_review > 0 or pos_out_meta_review > 0 or \
+       neg_in_meta_review > 0 or neg_out_meta_review > 0:
+        participant.meta_review_add_activity(participant)
```
Comment (li-boxuan) [inline gamification/process/update.py]: I don't think this would work as participant does not have `meta_review_add_activity` method.

**Comment by shikharvaish28** [comment]
By mistake, I clicked on ```Update``` and it caused the issue but it's now fixed.

File: .vscode/settings.json [hunk 15]
```diff
@@ -0,0 +1,3 @@
+{
+    "python.linting.enabled": false
+}
```
Comment (sks444) [inline .vscode/settings.json]: Please do not push this file.

File: gamification/process/update.py [hunk 16]
```diff
@@ -135,6 +136,33 @@ def get_participant_objects():
...
+
+
+def update_participants_data_with_meta_review(participant):
+    """
+    Update participants based on meta-review
+
+    This method updates every participant based on the meta-review
+    received or given. If the meta-review is completed then it calls
+    'Completed Meta-Review' activity.
+    """
```
Comment (sks444) [inline gamification/process/update.py]: First check if this participant is in the [active_newcomer_list](https://github.com/coala/community/blob/da5ada362c8fc3ae2cb179739a4babbfad1e7d1f/data/newcomers.py#L9), as gamification only work for those participant.
```
if participant.login in active_newcomers:
     ....
```

File: gamification/process/update.py [hunk 17]
```diff
@@ -135,6 +136,33 @@ def get_participant_objects():
...
+    neg_out_meta_review = participant.neg_out
+    if pos_in_meta_review > 0 or pos_out_meta_review > 0 or \
+       neg_in_meta_review > 0 or neg_out_meta_review > 0:
+        """
+        Add a meta-review activity to the participant
+        """
+        activity = Activity.object.create(
+            name=participant,
+            activity='Completed Meta-Review')
+        participant.activities.add(activity)
```
Comment (sks444) [inline gamification/process/update.py]: Now, this participant is a meta-review object.
You have to get a gamification participant first.
```
from gamification.models import Participant
gam_participant = Participant.objects.get_or_create(username=participant.login)
```

File: gamification/process/update.py [hunk 18]
```diff
@@ -135,6 +136,33 @@ def get_participant_objects():
...
+    pos_in_meta_review = participant.pos_in
+    pos_out_meta_review = participant.pos_out
+    neg_in_meta_review = participant.neg_in
+    neg_out_meta_review = participant.neg_out
+    if pos_in_meta_review > 0 or pos_out_meta_review > 0 or \
+       neg_in_meta_review > 0 or neg_out_meta_review > 0:
+        """
+        Add a meta-review activity to the participant
+        """
+        activity = Activity.object.create(
```
Comment (sks444) [inline gamification/process/update.py]: Try to use [add_activity](https://github.com/coala/community/blob/da5ada362c8fc3ae2cb179739a4babbfad1e7d1f/gamification/models.py#L135) method.

File: gamification/process/update.py (hunk 16 above)
Comment (shikharvaish28) [reply gamification/process/update.py]: Thanks!

File: gamification/process/update.py [hunk 19]
```diff
@@ -135,6 +136,33 @@ def get_participant_objects():
...
+
+    This method updates every participant based on the meta-review
+    received or given. If the meta-review is completed then it calls
+    'Completed Meta-Review' activity.
+    """
+    pos_in_meta_review = participant.pos_in
+    pos_out_meta_review = participant.pos_out
+    neg_in_meta_review = participant.neg_in
+    neg_out_meta_review = participant.neg_out
+    if pos_in_meta_review > 0 or pos_out_meta_review > 0 or \
```
[the remark this answers is not in the document]
Comment (li-boxuan) [reply gamification/process/update.py]: Avoid using `\`. If conditions are too long, use `(`, `)`.
```python
if (xxx or xxx or
        xxx or xxx):
```

**Comment by li-boxuan** [comment]
It would be much better if you can add some tests, e.g. creating two meta-review participants, one of whom has done a meta-review while the other one hasn't. Then obviously the former one completes the meta-review activity while the latter one doesn't complete.

File: gamification/process/update.py (hunk 16 above)
Comment (shikharvaish28) [reply gamification/process/update.py]: While adding the tests, I'm receiving this traceback 
``` 
    def update_participants_data_with_meta_review(meta_review_participant):
        """
        Update participants based on meta-review

        This method updates every participant based on the meta-review
        received or given. If the meta-review is completed then it calls
        'Did a meta-review or received a meta-review' activity.
        """
>       if meta_review_participant.login in active_newcomers:
E       AttributeError: 'Participant' object has no attribute 'login'

gamification/process/update.py:152: AttributeError
```

@sks444

Comment (li-boxuan) [reply gamification/process/update.py]: @shikharvaish28 It's hard to tell the reason unless you provide all code, but I guess it's because the `meta_review_participant` you are using here is NOT a meta-review participant, but a gamification participant, which does not have attribute `login`.

Comment (shikharvaish28) [reply gamification/process/update.py]: @li-boxuan @sks444 Please have a look!

File: gamification/process/update.py.orig [hunk 20]
```diff
@@ -0,0 +1,172 @@
+import logging
```
Comment (li-boxuan) [inline gamification/process/update.py.orig]: Remove this file.

File: gamification/tests/test_models.py [hunk 21]
```diff
@@ -341,3 +345,18 @@ def test_add_badges_method(self):
 
         # After applying add_badge method
         self.assertEquals(test_participant.badges.count(), 0)
+
+    def test_update_participants_data_with_meta_review(self):
+        # Test the participant who has completed the meta-review activity
+        Participant.objects.create(username='shikharvaish28')
+        test_meta_review_participant = Participant.objects.get(
+                                       username='shikharvaish28')
```
Comment (li-boxuan) [inline gamification/tests/test_models.py]: This is wrong. You are mixing up gamification participants and meta-review participants. You also need to import `Participant` from `meta-review` app.

File: gamification/tests/test_models.py [hunk 22]
```diff
@@ -341,3 +347,18 @@ def test_add_badges_method(self):
 
         # After applying add_badge method
         self.assertEquals(test_participant.badges.count(), 0)
+
+    def test_update_participants_data_with_meta_review(self):
+        # Test the participant who has completed the meta-review activity
+        meta_review_participant.objects.create(login='test2')
+        test_meta_review_participant = meta_review_participant.objects.get(
```
Comment (shikharvaish28) [inline gamification/tests/test_models.py]: I have an issue here. Here's what I try to do:

1. Create two meta-review participants. One with meta-review completed while the other has not.
2. Pass both of them through ```update_participants_data_with_meta_review()``` and update.
3. The one with meta-review completed would have added one activity unlike the other.


Now the issue is that when I'm creating the user through ```meta_review_participant```, I'm unable to get the activities count since it's a part of the ```gamification```.

Comment (li-boxuan) [reply gamification/tests/test_models.py]: Besides creating a meta-review participant, you also need to create a **corresponding** gamification participant.

Comment (shikharvaish28) [reply gamification/tests/test_models.py]: I tried to create a ```gamification``` participant with the same name as that of ```meta-review``` one. Now, the problem is how to link them. 
Though the activity is being added in ```meta-review```, how can I update it in ```gamification``` because in gamification, it's still showing activity count as 0 which in ideal situations should be 0?

File: gamification/process/update.py [hunk 23]
```diff
@@ -135,6 +137,37 @@ def get_participant_objects():
...
+            meta_review_participant.neg_in > 0 or
+                meta_review_participant.neg_out > 0):
+            """
+            Add a meta-review activity to the participant
+            """
+            name = meta_review_participant
+            activity = 'Did a meta-review or received a meta-review'
+            performed_at = Activity.performed_at
+            updated_at = Activity.updated_at
+            Participant.add_activity(name,
```
Comment (li-boxuan) [inline gamification/process/update.py]: Seems this is wrong. You should find the corresponding `gamification participant`. And do sth like `gamification_participant.add_activity()` instead of `Participant.add_activity`.

File: gamification/tests/test_models.py (hunk 22 above)
Comment (li-boxuan) [reply gamification/tests/test_models.py]: > Though the activity is being added in meta-review

You only need to play around with meta-review app to fetch some data you need. All the other stuff are related to `gamification`. `Activity` has nothing to do with `meta-review participant`.

File: gamification/process/update.py [hunk 24]
```diff
@@ -135,6 +137,39 @@ def get_participant_objects():
...
+
+def update_participants_data_with_meta_review(meta_review_participant):
+    """
+    Update participants based on meta-review
+
+    This method updates every participant based on the meta-review
+    received or given. If the meta-review is completed then it calls
+    'Did a meta-review or received a meta-review' activity.
+    """
+    if meta_review_participant.login == active_newcomers:
```
Comment (jayvdb) [inline gamification/process/update.py]: the tests say that line 153 - 166 here are not being tested at all, which means in the tests, this is always False. 

```console
gamification/process/update.py               69      8     19      1    85%   153-166, 152->153
```

And that is because this line is ridiculously wrong.  print out each side of this `==` and study the printout.

Comment (li-boxuan) [reply gamification/process/update.py]: Add print function, and use `.ci/build.sh` to see the output

Comment (shikharvaish28) [reply gamification/process/update.py]: Tried the same but I'm receiving this traceback:

```
(test_env) shikharvaish@Shikhars-MacBook-Air /Coding/GSoC/community (meta-review)
$ .ci/build.sh
+ rm db.sqlite3
rm: db.sqlite3: No such file or directory
+ true
+ rm -rf private _site public
+ mkdir private _site public
+ META_REVIEW_DATA=meta_review.json
+ EXPORTED_DATA='static/tasks.yaml static/instances.yaml static/meta_review.json'
+ ISSUES_JSON=issues.json
+ python manage.py fetch_deployed_data --allow-failure _site static/tasks.yaml static/instances.yaml static/meta_review.json
Traceback (most recent call last):
  File "manage.py", line 24, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/shikharvaish/anaconda3/envs/test_env/lib/python3.6/site-packages/django/core/management/__init__.py", line 367, in execute_from_command_line
    utility.execute()
  File "/Users/shikharvaish/anaconda3/envs/test_env/lib/python3.6/site-packages/django/core/management/__init__.py", line 359, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/shikharvaish/anaconda3/envs/test_env/lib/python3.6/site-packages/django/core/management/base.py", line 294, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/shikharvaish/anaconda3/envs/test_env/lib/python3.6/site-packages/django/core/management/base.py", line 345, in execute
    output = self.handle(*args, **options)
  File "/Coding/GSoC/community/data/management/commands/fetch_deployed_data.py", line 35, in handle
    upstream_deploy_url = get_upstream_deploy_url(repo_name, hoster)
  File "/Coding/GSoC/community/community/git.py", line 191, in get_upstream_deploy_url
    repo = get_upstream_repo()
  File "/Coding/GSoC/community/community/git.py", line 153, in get_upstream_repo
    parent = get_parent_repo(remote)
  File "/Coding/GSoC/community/community/git.py", line 119, in get_parent_repo
    repo = get_irepo(url)
  File "/Coding/GSoC/community/community/git.py", line 108, in get_irepo
    hoster = get_ihoster(url)
  File "/Coding/GSoC/community/community/git.py", line 87, in get_ihoster
    if url.resource == 'github.com':
AttributeError: 'list' object has no attribute 'resource'
```

File: gamification/process/update.py [hunk 25]
```diff
@@ -135,6 +137,42 @@ def get_participant_objects():
...
+            meta_review_participant.neg_in > 0 or
+                meta_review_participant.neg_out > 0):
+            """
+            Add a meta-review activity to the participant
+            """
+            activity = 'Did a meta-review or received a meta-review'
+            created_at = reaction_name.created_at
+            updated_at = '2017-08-24 05:59:31+00:00'
+            gamification_participant.add_activity(
+                                                0,
```
Comment (jayvdb) [inline gamification/process/update.py]: horrible indentation. find a nicer way.

File: gamification/process/update.py [hunk 26]
```diff
@@ -135,6 +137,42 @@ def get_participant_objects():
...
+    in the active newcomers list then it will check if the meta-review
+    is complete or not and update the activity accordingly.
+    """
+    active_newcomers_list = active_newcomers()
+    if meta_review_participant.login in active_newcomers_list:
+        gamification_participant = Participant.objects.get(
+            username=meta_review_participant.login)
+        reaction_name = Reaction.objects.get(id=meta_review_participant.login)
+        if (meta_review_participant.pos_in > 0 or
+            meta_review_participant.pos_out > 0 or
```
Comment (jayvdb) [inline gamification/process/update.py]: this violates pep8

File: gamification/process/update.py [hunk 27]
```diff
@@ -135,6 +137,38 @@ def get_participant_objects():
...
+                meta_review_participant.neg_in > 0 or
+                meta_review_participant.neg_out > 0):
+            """
+            Add a meta-review activity to the participant
+            """
+            activity = 'Did a meta-review or received a meta-review'
+            created_at = reaction_name.created_at
+            updated_at = None
+            gamification_participant.add_activity(0, activity,
+                                                  created_at, updated_at,)
```
Comment (jayvdb) [inline gamification/process/update.py]: a `,` before a `)` is completely useless.

the concept is called trailing comma , because it adds a comma and the end of the line -- the absolute end of the line, with nothing after it.

but also, use hanging indent for these args, dont put the args on the same line as the function name.  that isnt nice when the function name is long, as the args need to be squished on the RHS of the page.

File: gamification/process/update.py [hunk 28]
```diff
@@ -135,6 +137,38 @@ def get_participant_objects():
...
+        gamification_participant = Participant.objects.get(
+            username=meta_review_participant.login)
+        reaction_name = Reaction.objects.get(id=meta_review_participant.login)
+        if (meta_review_participant.pos_in > 0 or
+                meta_review_participant.pos_out > 0 or
+                meta_review_participant.neg_in > 0 or
+                meta_review_participant.neg_out > 0):
+            """
+            Add a meta-review activity to the participant
+            """
```
Comment (li-boxuan) [inline gamification/process/update.py]: Use `#` for comment here since it's single line anyway.

File: gamification/process/update.py [hunk 29]
```diff
@@ -135,6 +137,38 @@ def get_participant_objects():
...
+            activity = 'Did a meta-review or received a meta-review'
+            created_at = reaction_name.created_at
+            updated_at = None
+            gamification_participant.add_activity(0, activity, created_at,
+                                                  updated_at)
```
Comment (li-boxuan) [inline gamification/process/update.py]: I am not 100% sure here, but I suspect this would add repetitive activities to newcomers, because you don't check whether the newcomer has already completed a meta-review activity.

Comment (li-boxuan) [reply gamification/process/update.py]: Also add a testcase to verify that ;)
especially if you believe I am wrong here.

File: gamification/tests/test_models.py [hunk 30]
```diff
@@ -341,3 +349,48 @@ def test_add_badges_method(self):
 
         # After applying add_badge method
         self.assertEquals(test_participant.badges.count(), 0)
+
+    def test_update_participants_data_with_meta_review(self):
```
Comment (li-boxuan) [inline gamification/tests/test_models.py]: Can you split this long test case to several smaller test cases?

Comment (shikharvaish28) [reply gamification/tests/test_models.py]: Can you please throw some light on how you want me to split it? It's because all of the 3 participants here are for the same stuff so it's becoming difficult to identify which form do I split it in. @li-boxuan

File: gamification/process/update.py (hunk 29 above)
Comment (shikharvaish28) [reply gamification/process/update.py]: > I suspect this would add repetitive activities to newcomers

Sounds logical but in order to do this, I will have to add in a flag to ensure if the meta-review has been completed. 
Or should I create another list for the participants who have completed the meta-review. While passing on a participant's name, we can check if he's in the list or not. In case it is, we can break the statement.

Comment (li-boxuan) [reply gamification/process/update.py]: Maybe before adding a new activity, check if the user has already finished a meta-review activity. I’m not sure how gamification app avoids other duplicated activities, maybe you can learn from them.

File: gamification/process/update.py [hunk 31]
```diff
@@ -135,6 +137,42 @@ def get_participant_objects():
...
+        if (meta_review_participant.pos_in > 0 or
+                meta_review_participant.pos_out > 0 or
+                meta_review_participant.neg_in > 0 or
+                meta_review_participant.neg_out > 0):
+            # Add a meta-review activity to the participant
+            activity = 'Did a meta-review or received a meta-review'
+            created_at = reaction_name.created_at
+            updated_at = None
+            gamification_participant.add_activity(0, activity, created_at,
+                                                  updated_at)
```
Comment (sks444) [inline gamification/process/update.py]: Maybe add an `info level` log here saying that `this participant(get the username of the participant with gamification_participant.username ) has performed meta-review activity`.

File: gamification/process/update.py [hunk 32]
```diff
@@ -135,6 +137,42 @@ def get_participant_objects():
...
+    in the active newcomers list then it will check if the meta-review
+    is complete or not and update the activity accordingly.
+    """
+    active_newcomers_list = active_newcomers()
+    if (meta_review_participant.login in active_newcomers_list and
+            meta_review_participant.login not in meta_review_completed_list):
+        # Get the corresponding gamification participant
+        gamification_participant = Participant.objects.get(
+            username=meta_review_participant.login)
+        reaction_name = Reaction.objects.get(id=meta_review_participant.login)
```
[the remark this answers is not in the document]
Comment (shikharvaish28) [reply gamification/process/update.py]: ```Reaction.id``` and ```meta_review_participant.login```  are the name of the participant. (I've cross checked it by printing both of them.

File: gamification/process/update.py [hunk 33]
```diff
@@ -135,6 +137,38 @@ def get_participant_objects():
...
+def get_meta_review_participant_objects():
+    return meta_review_participants.objects.all()
+
+
+def update_participants_data_with_meta_review(meta_review_participant):
+    """
+    Update participants based on meta-review
+
+    This method updates every participant based on the meta-review
+    received or given. If the gamification participant is
```
[the remark this answers is not in the document]
Comment (shikharvaish28) [reply gamification/process/update.py]: I've done this already [here](https://github.com/coala/community/pull/190/files#diff-49eb4e29b7797643ccd41983aeb4c575R157) :)

[the remark this answers is not in the document]
Comment (sks444) [reply gamification/process/update.py]: No, I meant the sentence should be `If the meta-review participant is in..`?

File: gamification/process/update.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (li-boxuan) [reply gamification/process/update.py]: Really? How did you cross check that? That would be super weird.

[the remark this answers is not in the document]
Comment (shikharvaish28) [reply gamification/process/update.py]: In ```test_model.py```, I printed both of them separately and for seeing the output, ran ```pytest -s```

**Comment by li-boxuan** [comment]
I cannot see anyone who has finished `meta-review` activity on https://deploy-preview-190--coala-community.netlify.com/gamification/

**Comment by shikharvaish28** [comment]
> I cannot see anyone who has finished `meta-review` activity on https://deploy-preview-190--coala-community.netlify.com/gamification/

I'm myself able to understand why is it happening because the tests are running perfectly fine.

File: gamification/process/update.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (shikharvaish28) [reply gamification/process/update.py]: The ```meta_review_participant.login``` [output](https://pastebin.com/LEGhK3tS) and ```reaction``` [output](https://pastebin.com/6X6y5Kfx). 

The ```.ci/build.sh``` [output](https://pastebin.com/J9m8uBnt)

File: gamification/tests/test_models.py [hunk 34]
```diff
@@ -341,3 +349,64 @@ def test_add_badges_method(self):
...
         self.assertEquals(test_participant.badges.count(), 0)
+
+    def test_update_participants_data_with_meta_review_complete(self):
+        # Test the participant who has completed the meta-review activity
+        active_newcomers_list = active_newcomers()
+        active_newcomers_list.extend(['test2', 'test3'])
+        # meta-review participant
+        meta_review_participant.objects.create(login='test2')
+        test_meta_review_participant = meta_review_participant.objects.get(
+                                       login='test2')
```
Comment (li-boxuan) [inline gamification/tests/test_models.py]: Unnecessary. `meta_review_participant.objects.create(login='test2')` will return the instance.

File: gamification/tests/test_models.py [hunk 35]
```diff
@@ -341,3 +349,64 @@ def test_add_badges_method(self):
...
+
+        Reaction.objects.create(id='test2')
```
Comment (li-boxuan) [inline gamification/tests/test_models.py]: Well, this is definitely inappropriate and prone to error. You are creating a reaction with id also equals to `test2`, but the participant you created is also with id `test2`.

File: gamification/process/update.py [hunk 36]
```diff
@@ -135,6 +137,45 @@ def get_participant_objects():
...
+    in the active newcomers list then it will check if the meta-review
+    is complete or not and update the activity accordingly.
+    """
+    active_newcomers_list = active_newcomers()
+    if (meta_review_participant.login in active_newcomers_list and
+            meta_review_participant.login not in meta_review_completed_list):
+        # Get the corresponding gamification participant
+        gamification_participant = Participant.objects.get(
+            username=meta_review_participant.login)
+        reaction = Reaction.objects.get(id=meta_review_participant.login)
```
Comment (li-boxuan) [inline gamification/process/update.py]: This is definitely wrong. I checked the output you print out https://pastebin.com/6X6y5Kfx. You have the illusion that this works, because your test is wrong. In your test, both participant and reaction share the same id (primary_key), which gives you the wrong belief that you are correct.

That's why I said you should run `.ci/build.sh` and show the log to prove this works. You can use that for debugging, but you don't need to prove anymore as I am pretty confident now this is wrong.

Comment (li-boxuan) [reply gamification/process/update.py]: First you need to fix the test case. See https://github.com/coala/community/pull/190#discussion_r229155532 so that your test won't pass (which is good coz now you know your code is wrong).

Then read the definition of `Reaction` model. Understand the relationships between those models.

Comment (shikharvaish28) [reply gamification/process/update.py]: In order to test things, I used my user handle ```shikharvaish28```. Ideally, I should be able to get  ```reaction``` using ```id``` which is the primary key of ```Reaction``` model.

```python
test_meta_review_participant = meta_review_participant.objects.get(login='shikharvaish28')
reaction = Reaction.objects.get(id=test_meta_review_participant.login)
``` 
Instead, I am getting ```DoesNotExist: Reaction matching query does not exist.``` Can you please tell me why is it so?

Comment (li-boxuan) [reply gamification/process/update.py]: > Since login is the primary key of Participant, I am expecting it to be linked in someway to id of Reaction.

Your assumption is wrong. Django does not work like that.

I strongly recommend you to read https://docs.djangoproject.com/en/2.1/topics/db/queries/, esepecially understand how Foreign Key works.

**Comment by shikharvaish28** [comment]
@li-boxuan I'm closing this PR since I'm unable to solve this. I would prefer to do some other issues as of now. Kindly unassign me from this one so that this is free for someone else to take up.
